src/lib.rs

```diff
@@ -0,0 +1,219 @@
+extern crate regex;
+
+use regex::Regex;
+use std::collections::HashMap;
+
+#[derive(Debug, PartialEq)]
+pub enum IdentType {
+    Space,
+    Tab,
+    None
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Indent {
+    amount: usize,
+    indent: String,
+    ident_type: IdentType
+}
+
+fn ensure_indent_isize(x: usize) -> isize {
+    if x > std::isize::MAX as usize {
+        panic!("Indent greater then std::isize::MAX");
+    } else {
+        x as isize
+    }
+}
+
+fn get_most_used(indents: &HashMap<isize, (isize, isize)>) -> usize {
+    let mut result = 0;
+    let mut max_used = 0;
+    let mut max_weight = 0;
+
+    for (&key, &(u, w)) in indents.iter() {
+        if u > max_used || (u == max_used && w > max_weight) {
+            max_used = u;
+            max_weight = w;
+            result = key;
+        }
+    }
+
+    if result < 0 {
+        panic!("detect-indent::get_most_used cannot return a negative");
+    }
+
+    result as usize
+}
+
+pub fn detect_indent(string: &str) -> Indent {
+    let mut spaces = 0;
+    let mut tabs   = 0;
+    let indent_regex = Regex::new(r"^(?:( )+|\t+)").unwrap();
+    let mut indents: HashMap<isize, (isize, isize)> = HashMap::new();
+
+    let mut prev = 0;
+    let mut current :Option<isize>= None;
+    let mut key;
+
+    for line in string.lines() {
+        if line.is_empty() { continue; }
+        let mut indent = 0;
+
+        match indent_regex.captures(line) {
+            Some(captures) => {
+                captures.get(0).map( |capture| {
+                    let string = capture.as_str();
+                    indent = string.len();
+
+                    match string.chars().next().unwrap() {
+                        ' ' => spaces +=1,
+                        _ => tabs +=1,
+                    }
+                });
+            }
+
+            None => indent = 0
+        }
+
+        let iindent = ensure_indent_isize(indent);
```

Comment:
So this is really an assertion about the correctness of your library, and your consumers won't want it to panic. So I don't think the word "ensure" is right -- I'd typically use that for validating external input. I think you want to delete your `ensure_indent_isize` function and replace this line with:

```rust
assert!(indent <= (std::isize::MAX as usize), "indent greater than std::isize::MAX");
let iindent = indent as usize;
```